### ISSUE TITLE
systemd: mount units fail with "Mount process finished, but there is no mount."

### ISSUE THREAD

**Issue by bboozzoo** [systemd/systemd#10872]
**systemd version the issue has been seen with**
> 239, 238, 237

**Used distribution**
 > Arch
> Ubuntu 18.10, 18.04
> Fedora 28 & 29

**Expected behaviour you didn't see**
 > mounts work

**Unexpected behaviour you saw**
 > some mount units fail with
```
kernel: EXT4-fs (loop1): mounted filesystem with ordered data mode. Opts: (null)
systemd[1]: tmp-mounttest-mount-15.mount: Mount process finished, but there is no mount.                                                                                                                                                                                                                             
systemd[1]: tmp-mounttest-mount-15.mount: Failed with result 'protocol'.                                                                                                                                                                                                                                             
systemd[1]: Failed to mount mount unit for test 15.
```

**Steps to reproduce the problem**
Grab the reproducer script: https://gist.github.com/bboozzoo/d4b142229b1915ef7cc0cf8593599ad9/828d716e484a39da11987b2dc38da86434d1f89f

We have been tracking the problem in snapd for a while in https://forum.snapcraft.io/t/unexplained-mount-failure-protocol-error-what-we-know-so-far/5682 and it started appearing around April/May 2018. It randomly reproduces in the CI runs on distros with recent(-ish) systemd while installing snaps. What `snapd` does is in short: generate a mount unit for the snap, drop it under `/etc/systemd/system`, call `systemctl daemon-reload`, and later `systemctl start <...>.mount`. The last step randomly fails. The journal message is what is provided above.

The reproducer script was used to explore some possible ideas on how it fails. So far, only the variant when the thing happens is daemon-reload interleaved with start/stop of mount units. On my machine it fails reliably in 1-2 loop iterations. Other variants that were explored and failed to reproduce: loading mount units before and doing start/stop, using mount directly, calling systemd-mount (this one failed in a precular, but unrelated way).

I also tried Fedora 28/29 cloud images and Ubuntu 18.10/18.04 cloud image with similar results.

Edit: added Ubuntu 18.04

**Comment by poettering** [comment]
If you replace the mount binary with a shell script that invokes the original mount binary but after it checks /proc/self/mountinfo to see if the mount is actually established, what do you see?

**Comment by bboozzoo** [comment]
Sorry late feedback. This is what I get:
```
Nov 29 13:50:36 localhost systemd[1]: Reloading.
Nov 29 13:50:36 localhost systemd[1]: Mounting mount unit for test 5...
Nov 29 13:50:36 localhost kernel: EXT4-fs (loop0): mounted filesystem with ordered data mode. Opts: (null)
Nov 29 13:50:36 localhost mount[4082]: 299 90 7:0 / /tmp/mounttest/mount/5 rw,nodev,relatime shared:160 - ext4 /dev/loop0 rw,seclabel
Nov 29 13:50:36 localhost systemd[1]: Mounted mount unit for test 5.
Nov 29 13:50:36 localhost systemd[1]: Reloading.
Nov 29 13:50:36 localhost systemd[1]: Mounting mount unit for test 10...
Nov 29 13:50:36 localhost systemd[1]: Reloading.
Nov 29 13:50:36 localhost kernel: EXT4-fs (loop1): mounted filesystem with ordered data mode. Opts: (null)
Nov 29 13:50:37 localhost mount[4114]: 308 90 7:1 / /tmp/mounttest/mount/10 rw,nodev,relatime shared:165 - ext4 /dev/loop1 rw,seclabel
Nov 29 13:50:37 localhost systemd[1]: tmp-mounttest-mount-10.mount: Mount process finished, but there is no mount.
Nov 29 13:50:37 localhost systemd[1]: tmp-mounttest-mount-10.mount: Failed with result 'protocol'.
Nov 29 13:50:37 localhost systemd[1]: Failed to mount mount unit for test 10.
Nov 29 13:50:37 localhost systemd[1]: Reloading.
Nov 29 13:50:37 localhost systemd[1]: Reloading.
Nov 29 13:50:37 localhost systemd[1]: Mounting mount unit for test 16...
Nov 29 13:50:37 localhost systemd[1]: Mounting mount unit for test 4...
Nov 29 13:50:37 localhost systemd[1]: Mounting mount unit for test 15...
Nov 29 13:50:37 localhost kernel: EXT4-fs (loop2): mounted filesystem with ordered data mode. Opts: (null)
Nov 29 13:50:37 localhost kernel: EXT4-fs (loop3): mounted filesystem with ordered data mode. Opts: (null)
Nov 29 13:50:37 localhost kernel: EXT4-fs (loop4): mounted filesystem with ordered data mode. Opts: (null)
Nov 29 13:50:37 localhost mount[4185]: 317 90 7:2 / /tmp/mounttest/mount/16 rw,nodev,relatime shared:170 - ext4 /dev/loop2 rw,seclabel
Nov 29 13:50:37 localhost systemd[1]: Mounted mount unit for test 16.                                                                                                                                                                                                                                                                                      
Nov 29 13:50:37 localhost mount[4186]: 318 90 7:3 / /tmp/mounttest/mount/4 rw,nodev,relatime shared:175 - ext4 /dev/loop3 rw,seclabel
Nov 29 13:50:37 localhost systemd[1]: Mounted mount unit for test 4.
Nov 29 13:50:37 localhost mount[4189]: 319 90 7:4 / /tmp/mounttest/mount/15 rw,nodev,relatime shared:180 - ext4 /dev/loop4 rw,seclabel
Nov 29 13:50:37 localhost systemd[1]: Mounted mount unit for test 15.
Nov 29 13:50:37 localhost systemd[1]: Reloading.
Nov 29 13:50:37 localhost systemd[1]: Reloading.
```

For the record, this is the wrapper I used:
```
#!/bin/bash
marg=
for arg in "$@"; do
    if [[ "$arg" = /tmp/mounttest/mount* ]]; then
       marg=$arg
       break
    fi
done
/usr/bin/mount.orig "$@"
status=$?
if [[ -n "$marg" ]]; then
   grep "$marg" /proc/self/mountinfo
fi
exit $status
```

**Comment by poettering** [comment]
I am pretty sure #10980 will fix this one too. Any chance you can give it a whirl?

**Comment by bboozzoo** [comment]
Running with revision ca7d7db84 from #10980:

```
Dec 03 15:38:55 archlinux kernel: EXT4-fs (loop6): mounted filesystem with ordered data mode. Opts: (null)                     
Dec 03 15:38:55 archlinux mount[3664]: 421 48 7:6 / /tmp/mounttest/mount/12 rw,nodev,relatime shared:242 - ext4 /dev/loop6 rw
Dec 03 15:38:55 archlinux systemd[1]: -.slice: Failed to set 'cpu.cfs_period_us' attribute on '/' to '100000': Invalid argument
Dec 03 15:38:55 archlinux systemd[1]: -.slice: Failed to set 'cpu.cfs_quota_us' attribute on '/' to '-1': Invalid argument   
Dec 03 15:38:55 archlinux systemd[1]: -.slice: Failed to set 'memory.limit_in_bytes' attribute on '/' to '-1': Invalid argument
Dec 03 15:38:55 archlinux systemd[1]: tmp-mounttest-mount-12.mount: Mount process finished, but there is no mount.           
Dec 03 15:38:55 archlinux systemd[1]: tmp-mounttest-mount-12.mount: Failed with result 'protocol'.                           
Dec 03 15:38:55 archlinux systemd[1]: Failed to mount mount unit for test 12.                                                
Dec 03 15:38:55 archlinux systemd[1]: Reloading.                                                                               
```

FWIW I noticed that the script was hitting another sanity check, which is run in the same process, after `systemctl start`, but before `systemctl stop`:
```
verify() {
    local i
    i="$1"

    if ! grep -q "$PREFIX/mount/$i " /proc/self/mountinfo; then
        echo "disk-$i missing"
        return 1
    fi
   ...
}
```

The log says:
```
Dec 03 15:38:58 archlinux mount[3989]: 384 48 7:1 / /tmp/mounttest/mount/7 rw,nodev,relatime shared:232 - ext4 /dev/loop1 rw
Dec 03 15:38:58 archlinux systemd[1]: Mounted mount unit for test 7.
Dec 03 15:38:58 archlinux kernel: EXT4-fs (loop2): mounted filesystem with ordered data mode. Opts: (null)
Dec 03 15:38:58 archlinux mount[3994]: 411 48 7:2 / /tmp/mounttest/mount/10 rw,nodev,relatime shared:247 - ext4 /dev/loop2 rw
Dec 03 15:38:58 archlinux systemd[1]: Unmounting mount unit for test 10...
Dec 03 15:38:58 archlinux systemd[1]: Unmounting mount unit for test 7...
Dec 03 15:38:58 archlinux systemd[271]: tmp-mounttest-mount-10.mount: Succeeded.
Dec 03 15:38:58 archlinux systemd[271]: tmp-mounttest-mount-7.mount: Succeeded.
Dec 03 15:38:58 archlinux systemd[1]: tmp-mounttest-mount-10.mount: Succeeded.
Dec 03 15:38:58 archlinux systemd[1]: Unmounted mount unit for test 10.
Dec 03 15:38:58 archlinux systemd[1]: tmp-mounttest-mount-7.mount: Succeeded.
Dec 03 15:38:58 archlinux systemd[1]: Unmounted mount unit for test 7.
```

**Comment by bboozzoo** [comment]
We were checking some potential workarounds in snapd. We've simplified the reproducer script to do just one mount, but many reloads in parallel and we were able to reproduce the problem, though it took much longer.

Now that #10980 is merged, would you like me to try with the latest master?

**Comment by myrkr** [comment]
I still have seen this problem of parallel execution of {{systemctl daemon-reload}} (after creating some new service unit) and parallel execution of mount units with the error message:

```
2019-03-04 14:00:52.640 init.scope: var-lib-machines-grade2.mount: Mount process finished, but there is no mount.
2019-03-04 14:00:52.640 init.scope: var-lib-machines-grade2.mount: Failed with result 'protocol'.
```

in version v241.

We work around this problem by trusting the successful execution of the mount command:

```
diff --git a/src/core/mount.c b/src/core/mount.c
index c31cad6b52..a8327fe355 100644
--- a/src/core/mount.c
+++ b/src/core/mount.c
@@ -1347,9 +1347,11 @@ static void mount_sigchld_event(Unit *u, pid_t pid, int code, int status) {
                         /* Either /bin/mount has an unexpected definition of success,
                          * or someone raced us and we lost. */
                         log_unit_warning(UNIT(m), "Mount process finished, but there is no mount.");
-                        f = MOUNT_FAILURE_PROTOCOL;
+                        log_unit_warning(UNIT(m), "But assuming success anyway");
+                        mount_enter_mounted(m, f);
+                } else {
+                        mount_enter_dead(m, f);
                 }
-                mount_enter_dead(m, f);
                 break;
 
         case MOUNT_MOUNTING_DONE:

```

Unfortunately, I have not yet been able to isolate a reproducer for this problem, it just fails from time to time when creating new systemd units (and reloading) and performing mounts as dependency for starting other service units.

**Comment by changyp6** [comment]
> I still have seen this problem of parallel execution of `systemctl daemon-reload` (after creating some new service unit) and parallel execution of mount units with the error message:
> 
> ```
> 2019-03-04 14:00:52.640 init.scope: var-lib-machines-grade2.mount: Mount process finished, but there is no mount.
> 2019-03-04 14:00:52.640 init.scope: var-lib-machines-grade2.mount: Failed with result 'protocol'.
> ```
> in version v241.
> 
> We work around this problem by trusting the successful execution of the mount command:
> 
> ```
> diff --git a/src/core/mount.c b/src/core/mount.c
> index c31cad6b52..a8327fe355 100644
> --- a/src/core/mount.c
> +++ b/src/core/mount.c
> @@ -1347,9 +1347,11 @@ static void mount_sigchld_event(Unit *u, pid_t pid, int code, int status) {
>                          /* Either /bin/mount has an unexpected definition of success,
>                           * or someone raced us and we lost. */
>                          log_unit_warning(UNIT(m), "Mount process finished, but there is no mount.");
> -                        f = MOUNT_FAILURE_PROTOCOL;
> +                        log_unit_warning(UNIT(m), "But assuming success anyway");
> +                        mount_enter_mounted(m, f);
> +                } else {
> +                        mount_enter_dead(m, f);
>                  }
> -                mount_enter_dead(m, f);
>                  break;
>  
>          case MOUNT_MOUNTING_DONE:
> ```
> Unfortunately, I have not yet been able to isolate a reproducer for this problem, it just fails from time to time when creating new systemd units (and reloading) and performing mounts as dependency for starting other service units.

I have this same unmounting tmp failed issue, when rebooting the system.
Even if I stopped all the processes that use /tmp, umouting tmp still reported.
I'll try your patch to see if it works.

**Comment by frispete** [comment]
I do see this issue with nfs mounts on openSUSE Tumbleweed, that carries 241.

```
$ systemctl --failed
  UNIT           LOAD   ACTIVE SUB    DESCRIPTION                                                                                                                                                  
● home-nfs.mount loaded failed failed /home/nfs                                                                                                                                                    
● video.mount    loaded failed failed /video                                                                                                                                                       
● work.mount     loaded failed failed /work         
```

```
$ grep nfs /etc/fstab
server:/home    /home/nfs  nfs   nfsvers=4,x-systemd.requires=/home,x-systemd.requires=nfs-mountd.service  0  0
server:/work    /work      nfs   nfsvers=4,x-systemd.requires=nfs-mountd.service  0  0
server:/video   /video     nfs   nfsvers=4,x-systemd.requires=nfs-mountd.service  0  0
```

```
syslog:
2019-04-14T15:29:22.925116+02:00 xrated systemd[1]: Started System Security Services Daemon.
2019-04-14T15:29:22.925227+02:00 xrated systemd[1]: home-nfs.mount: Mount process finished, but there is no mount.
2019-04-14T15:29:22.925289+02:00 xrated systemd[1]: home-nfs.mount: Failed with result 'protocol'.
2019-04-14T15:29:22.925610+02:00 xrated systemd[1]: Failed to mount /home/nfs.
2019-04-14T15:29:22.925818+02:00 xrated systemd[1]: Dependency failed for Remote File Systems.
2019-04-14T15:29:22.925866+02:00 xrated systemd[1]: remote-fs.target: Job remote-fs.target/start failed with result 'dependency'.
2019-04-14T15:29:22.925984+02:00 xrated systemd[1]: work.mount: Mount process finished, but there is no mount.
2019-04-14T15:29:22.926029+02:00 xrated systemd[1]: work.mount: Failed with result 'protocol'.
2019-04-14T15:29:22.926246+02:00 xrated systemd[1]: Failed to mount /work.
2019-04-14T15:29:22.926495+02:00 xrated systemd[1]: video.mount: Mount process finished, but there is no mount.
2019-04-14T15:29:22.926548+02:00 xrated systemd[1]: video.mount: Failed with result 'protocol'.
2019-04-14T15:29:22.926780+02:00 xrated systemd[1]: Failed to mount /video.
2019-04-14T15:29:22.927205+02:00 xrated systemd[1]: rpc-statd-notify.service: Succeeded.
2019-04-14T15:29:22.927493+02:00 xrated systemd[1]: Started Notify NFS peers of a restart.
2019-04-14T15:29:22.928432+02:00 xrated systemd[1]: Reached target User and Group Name Lookups.
2019-04-14T15:29:22.929360+02:00 xrated systemd[1]: Starting Permit User Sessions...
2019-04-14T15:29:22.930346+02:00 xrated systemd[1]: Starting Login Service...
2019-04-14T15:29:22.938858+02:00 xrated systemd[1]: Reloading.
```

**Comment by frispete** [comment]
Forgot to mention, the mounts are working fine, nevertheless.

**Comment by hcderaad** [comment]
Can confirm this same issue on recent openSUSE Tumbleweed releases with version:
Name           : systemd
Version        : 241-2.1
Arch           : x86_64
Vendor         : openSUSE

In this case it is an encrypted XFS partition. Mount fails on boot, yet is fully usable, exactly like @frispete reports. Any updates or possibly any testcases to run perhaps?

**Comment by poettering** [comment]
I prepped a proposal to fix this in #13097, ptal!

**Comment by bboozzoo** [comment]
Thanks for the fix. Unfortunately I can still reproduce the problem on Arch using the script I posted initially and the latest master of systemd and the stock Arch 5.2.1 kernel:

```
Jul 22 09:44:27 archlinux systemd[1]: systemd 242 running in system mode. (+PAM +AUDIT -SELINUX +IMA -APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ 
+LZ4 +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=unified)
Jul 22 09:44:27 archlinux systemd[1]: Detected virtualization kvm.
Jul 22 09:44:27 archlinux systemd[1]: Detected architecture x86-64.
Jul 22 09:44:27 archlinux systemd[1]: Initializing machine ID from random generator.
Jul 22 09:44:27 archlinux systemd[1]: initrd-switch-root.service: Succeeded.
Jul 22 09:44:27 archlinux systemd[1]: Stopped Switch Root.
...
Jul 22 10:04:24 archlinux systemd[1]: tmp-mounttest-mount-17.mount: Mount process finished, but there is no mount.
Jul 22 10:04:24 archlinux systemd[1]: tmp-mounttest-mount-17.mount: Failed with result 'protocol'.
Jul 22 10:04:24 archlinux systemd[1]: Failed to mount mount unit for test 17.
Jul 22 10:04:24 archlinux systemd[1]: tmp-mounttest-mount-2.mount: Mount process finished, but there is no mount.
Jul 22 10:04:24 archlinux systemd[1]: tmp-mounttest-mount-2.mount: Failed with result 'protocol'.
Jul 22 10:04:24 archlinux systemd[1]: Failed to mount mount unit for test 2.
```

I tried the Arch LTS kernel too (4.19.60-1-lts) and the problem is reproducible as well.

Let me know if I can do anything to help. The images were built using mkosi (adding some extra packages for the target, like sshd and such).

**Comment by keszybz** [comment]
@bboozzoo are you sure you are running the patched version?
> Jul 22 09:44:27 archlinux systemd[1]: systemd 242 running in system mode.

This should be something like `system 242-1498-g350804867dd`  or later if you installed directly from git.

**Comment by bboozzoo** [comment]
@keszybz not sure why the tagger did not run, anyways, the stock systemd only ran in initrd and logged something in the lines of `systemd 242.84-1 running in system mode.`

FWIW, the same version was logged when I built a Fedora image with mkosi.

**Comment by keszybz** [comment]
So... I tried your reproducer, and things indeed fail, but I think they fail not in systemd but in mount.
My logs are full of:
```
Jul 30 17:48:07 krowka systemd[1]: Mounting mount unit for test 8...
Jul 30 17:48:07 krowka kernel: print_req_error: I/O error, dev loop12, sector 0 flags 801
Jul 30 17:48:07 krowka kernel: print_req_error: I/O error, dev loop12, sector 0 flags 801
Jul 30 17:48:07 krowka systemd[1]: tmp-mounttest-mount-2.mount: Succeeded.
Jul 30 17:48:07 krowka systemd[1]: Unmounted mount unit for test 2.
Jul 30 17:48:07 krowka kernel: EXT4-fs (loop8): mounted filesystem with ordered data mode. Opts: (null)
Jul 30 17:48:07 krowka systemd[1]: Mounted mount unit for test 20.
Jul 30 17:48:07 krowka kernel: EXT4-fs (loop12): mounted filesystem with ordered data mode. Opts: (null)
Jul 30 17:48:07 krowka systemd[1]: tmp-mounttest-mount-4.mount: Mount process exited, code=killed, status=15/TERM
Jul 30 17:48:07 krowka systemd[1]: Failed unmounting mount unit for test 4.
```
and then `tmp-mounttest-mount-4.mount` is stopped:
```
+ systemctl stop tmp-mounttest-mount-2.mount
+ systemctl start tmp-mounttest-mount-4.mount
Job for tmp-mounttest-mount-4.mount canceled.
+ systemctl stop tmp-mounttest-mount-7.mount
```

The interesting part is that the I/O errors don't cause failure on their own. They happen quite often, and only sometimes the test fails. Neverthless, when the test fails, it seems to be always preceded by an I/O error.

I'm don't know who is sending the TERM signal, and why systemd considered the mount cancelled. 

My hypothesis is that the autoassignment of loop devices goes awry and this causes mounts to fail.

**Comment by bboozzoo** [comment]
We are also keeping an eye on some recent issues with 5.2 kernels around loopback mounts: https://lore.kernel.org/linux-block/f7e6717f-be46-1dfe-80ed-f85a18065c85@i-love.sakura.ne.jp/T/#m7b86e0c0f368ad6523c80b02bcd758d387a93a77 Perhaps it's related

**Comment by keszybz** [comment]
So... I changed the assignment script to use fixed device assignments, by adding:
```
Options=nodev,loop=/dev/loop$((i+6))
```
With this change, the mounts don't fail.

In fact, I'm disappointed to see that `mount` used the traditional "let's find an unused loop device" approach:
```console
$ sudo strace -eopenat,close  /bin/mount /tmp/mounttest/disk-8.img /tmp/mounttest/mount/8 -t ext4 -o nodev
...
openat(AT_FDCWD, "/sys/block", O_RDONLY|O_NONBLOCK|O_CLOEXEC|O_DIRECTORY) = 3
openat(AT_FDCWD, "/sys/dev/block/7:1", O_RDONLY|O_CLOEXEC) = 4
openat(4, "loop/offset", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
openat(AT_FDCWD, "/dev/loop1", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
close(4)                                = 0
openat(AT_FDCWD, "/sys/dev/block/7:6", O_RDONLY|O_CLOEXEC) = 4
openat(4, "loop/offset", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
openat(AT_FDCWD, "/dev/loop6", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
close(4)                                = 0
openat(AT_FDCWD, "/sys/dev/block/7:4", O_RDONLY|O_CLOEXEC) = 4
openat(4, "loop/offset", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
openat(AT_FDCWD, "/dev/loop4", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
close(4)                                = 0
openat(AT_FDCWD, "/sys/dev/block/7:2", O_RDONLY|O_CLOEXEC) = 4
openat(4, "loop/offset", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
openat(AT_FDCWD, "/dev/loop2", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
close(4)                                = 0
openat(AT_FDCWD, "/sys/dev/block/7:0", O_RDONLY|O_CLOEXEC) = 4
openat(4, "loop/offset", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
openat(AT_FDCWD, "/dev/loop0", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
close(4)                                = 0
openat(AT_FDCWD, "/sys/dev/block/7:5", O_RDONLY|O_CLOEXEC) = 4
openat(4, "loop/offset", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
openat(AT_FDCWD, "/dev/loop5", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
close(4)                                = 0
openat(AT_FDCWD, "/sys/dev/block/7:12", O_RDONLY|O_CLOEXEC) = 4
openat(4, "loop/offset", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
openat(AT_FDCWD, "/dev/loop12", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
close(4)                                = 0
openat(AT_FDCWD, "/sys/dev/block/7:3", O_RDONLY|O_CLOEXEC) = 4
openat(4, "loop/offset", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
openat(AT_FDCWD, "/dev/loop3", O_RDONLY|O_CLOEXEC) = 5
close(3)                                = 0
close(5)                                = 0
close(4)                                = 0
openat(AT_FDCWD, "/dev/loop-control", O_RDWR|O_CLOEXEC) = 3
close(3)                                = 0
openat(AT_FDCWD, "/tmp/mounttest/disk-8.img", O_RDWR|O_CLOEXEC) = 3
openat(AT_FDCWD, "/dev/loop7", O_RDWR|O_CLOEXEC) = 4
close(3)                                = 0
openat(AT_FDCWD, "/sys/dev/block/7:7", O_RDONLY|O_CLOEXEC) = 3
openat(3, "loop/autoclear", O_RDONLY|O_CLOEXEC) = 5
close(5)                                = 0
openat(3, "ro", O_RDONLY|O_CLOEXEC)     = 5
close(5)                                = 0
close(3)                                = 0
close(4)                                = 0
close(1)                                = 0
close(2)                                = 0
```
There seems to be no locking involved. With many concurrent attempts to do this, failure seems inevitable.

@karelzak comments?

**Comment by karelzak** [comment]
You're making conclusion based on incomplete strace and without code ;-) The libmount uses /dev/loop-contol to get unused free loop device of course.

I guess you have more loop devices, in this case libmount tries to verify that the backing file is not already associated with another loop device, because kernel does not care about it and it creates more loop devices for the same file and after that more independent FS instances are mounted. The result is FS damage.

The current code uses backing file devno and inode number for the verification and it's available only by LOOP_GET_STAT64 ioctl (so you see openat for the devices). 

The another possible way is to read backing filename from /sys and compare only paths rather than devno+inode, this optimization will save open+ioctl, the question is reliability as file path means nothing (due to namespaces, mounts overlaps, etc.).
```
# strace -eopenat,ioctl mount /root/file.img /mnt/test -t ext4 -o nodev
...
openat(AT_FDCWD, "/dev/loop0", O_RDONLY|O_CLOEXEC) = 3
ioctl(3, LOOP_GET_STATUS64, {lo_offset=0, lo_number=0, lo_flags=LO_FLAGS_AUTOCLEAR, lo_file_name="/root/file.img", ...}) = 0
...
mount: /mnt/test: /root/file.img is already mounted.
+++ exited with 1 +++
```
And it's more complicated, if you want to mount the same backing file on another place than libmount detects it and reuse already created loop device (if with the same options).
```
# mount /root/file.img /mnt/test/A -t ext4 -o nodev
# mount /root/file.img /mnt/test/A -t ext4 -o nodev
mount: /mnt/test/A: /root/file.img is already mounted.

#  mount /root/file.img /mnt/test/B -t ext4 -o nodev
# findmnt
TARGET                                SOURCE         FSTYPE          OPTIONS
...
├─/mnt/test/B                         /dev/loop0     ext4            rw,nodev,relatime
└─/mnt/test/A                         /dev/loop0     ext4            rw,nodev,relatime
```

**Comment by karelzak** [comment]
And note that the issue is also reported against NFS by @frispete.

**Comment by keszybz** [comment]
@karelzak thanks for the answer. I shouldn't have cast blame on mount so easily, sorry!
Nevertheless, it seems to not be working as it should.
@frispete's case is irrelevant, because it's with systemd-241, but we only care about systemd-243-rc1+, since the purported fix was just committed a few days ago). Please look at my results in the last few comments.

There are two issues:
1. kernel reports `print_req_error: I/O error, dev loopNN sector 0` occasionally.
2. mounts fail, and they *seem* to be failing in the `mount` directly. From the logs, it seems that systemd is simply calling `mount` and `umount` pairs correctly, and occasionally the mount will fail.

**Comment by hcderaad** [comment]
> 
> 
> Can confirm this same issue on recent openSUSE Tumbleweed releases with version:
> Name : systemd
> Version : 241-2.1
> Arch : x86_64
> Vendor : openSUSE
> 
> In this case it is an encrypted XFS partition. Mount fails on boot, yet is fully usable, exactly like @frispete reports. Any updates or possibly any testcases to run perhaps?

I can confirm the issue of mount errors using an encrypted XFS partition is no longer present in my setup in the following version, thanks for fixing this issue!

Repository     : openSUSE-Tumbleweed-Oss
Name           : systemd
Version        : 242-3.1
Arch           : x86_64
Vendor         : openSUSE

**Comment by keszybz** [comment]
To clarify this, I tried to reproduce this without systemd. So I boot a VM with `init=/bin/bash`, and I run `--variant=mount`, i.e. the variant that essentially does `mount -o nodev /tmp/loop/disk-$i.img /tmp/loop/mount/$i; sleep 1; verify $i; umount $i` with `$COUNT` such sequences running in parallell.  Without systemd in the picture, it *is* harder to reproduce, but the error still happens. I had to raise COUNT to 120 to see the issue, even though it occurs fairly reliably with COUNT=20 when systemd is running.

There are no visible errors, but after `mount` returns, the mount point is not listed in `/proc/self/mountinfo`. The question is: is this a valid expectation for it to be present?

**Comment by keszybz** [comment]
This is the simplified reproducer:
[reproduce3.sh.log](https://github.com/systemd/systemd/files/3524856/reproduce3.sh.log)
Run with `bash ./reproduce.sh`.
It fails with `disk-NN is missing` within a few interations in a rawhide VM with kernel-core-5.3.0-0.rc5.git0.1.fc32.x86_64, util-linux-2.34-2.fc31.x86_64.

**Comment by keszybz** [comment]
[strace-106.log](https://github.com/systemd/systemd/files/3524945/strace-106.log)
Stracing shows that `mount` syscall is called without any fuss, and `mount` exits. The mount is not  immediately present in /proc/self/mountinfo, but appears a moment later and remains there.
So this looks like a kernel issue, or a mismatched expectation.

In 350804867dbcc9b7ccabae1187d730d37e2d8a21, @poettering wrote:
> In theory, the kernel fires /proc/self/mountinfo changes off before returning from the mount() or umount() syscalls

Is this documented somewhere?

**Comment by keszybz** [comment]
strace of the `mount` and subsequent `cat /proc/self/mountinfo`.
[mount-38.mount.log](https://github.com/systemd/systemd/files/3525679/mount-38.mount.log)
[mount-38.cat.log](https://github.com/systemd/systemd/files/3525681/mount-38.cat.log)

**Comment by karelzak** [comment]
Conclusion from discussion with kernel VFS devels:

It seems kernel is not able to provide reliable mountinfo if userspace reads the file by more read() syscalls and another processes modify the mount table in the same time. The recommended solution is to minimize number of read()s and use poll() after the read() to check that the table has not been modified during the last read() -- I'll try to implement this into libmount next week, and kernel developers will try to improve kernel in next versions to make mountinfo more robust.

**Comment by poettering** [comment]
OK, then let's remove this from the v243 milestone, and assume for now that systemd is doing everything right as soon as libmount is fixed. we can revisit this in the v244 cycle, if this is really fixed then.

**Comment by karelzak** [comment]
@keszybz libmount in util-linux git tree contains the improvement as suggested by Al Viro (read()+poll()).

I didn't test it with systemd, but I have replaced "cat /proc/self/mountinfo" with "./test_mount_utils --read-procfs /proc/self/mountinfo" in your reproducer and it seems it works as expected. (The command ./test_mount_utils is a small libmount test binary.) 

The original test fails always, the new version pass, I have also tried  "COUNT=230" and it works too. If the COUNT= is bigger than it fails again (but maybe the problem is number of the available loopdevs nodes in this case).

It seems more interesting to use tmpfs rather loops (because tmpfs is without blockdev layer and udevd).
In this case it's possible to reproduce the problem with COUNT=10 (!) for the original version; the new version works with COUNT=300.

So, from my point of view it seems better now.

**Comment by keszybz** [comment]
Thanks. I'll try to test this with systemd (probably some time next week). But it sounds like this should fix the issue.

**Comment by bluca** [comment]
Hi,

Tried the patches and unfortunately I still see the issue (mount unit fails with 'protocol' error despite mount working) at least once with ~20 mount units being activated within a handful of seconds.

This is on an arm64 distro built based on Yocto (thud) running on Qemu, with kernel 4.19 LTS, util-linux 2.32.1 plus the  "libmount: improve mountinfo reliability" and "libmount; fix and improve read+poll mountinfo" patches, and systemd 242 plus the following patches:

fdeea3f4f1 Add fopen_unlocked() wrapper
41f6e627d7 Make fopen_temporary and fopen_temporary_label unlocked
13dcfe4661 shared/mount-util: convert to libmount
7d991d4818 mount-tool: use libmount to parse /proc/self/mountinfo
e2857b3d87 Add helper function for mnt_table_parse_{stream,mtab}
2f2d81d957 shared/mount-util: make sure utab is ignored in umount_recursive()
d34a40082d mount-util: bind_remount: avoid calling statvfs
fcd8e119c2 mount: simplify /proc/self/mountinfo handler
350804867d mount: rescan /proc/self/mountinfo before processing waitid() results
9ddaa3e459 mount: rename update_parameters_proc_self_mount_info() → update_parameters_proc_self_mountinfo()

Is there some other systemd patch that should be backported on top of those by any chance?

Thanks!

**Comment by bluca** [comment]
The reproducer script posted at https://github.com/systemd/systemd/issues/10872#issuecomment-523399087 fails after 61 mounts:

```
--- run 1 Mon Sep  2 16:23:34 UTC 2019
mount: /tmp/mounttest/mount/1: wrong fs type, bad option, bad superblock on /dev/loop2, missing codepage or helper program, or other error.
^Cinterrupted: run with --cleanup
-- failed units, skipping cleanup, use --cleanup manually
# losetup
NAME        SIZELIMIT OFFSET AUTOCLEAR RO BACK-FILE                  DIO LOG-SEC
/dev/loop1          0      0         1  0 /tmp/mounttest/disk-4.img    0     512
/dev/loop29         0      0         1  0 /tmp/mounttest/disk-31.img   0     512
/dev/loop57         0      0         1  0 /tmp/mounttest/disk-58.img   0     512
/dev/loop19         0      0         1  0 /tmp/mounttest/disk-19.img   0     512
/dev/loop47         0      0         1  0 /tmp/mounttest/disk-50.img   0     512
/dev/loop37         0      0         1  0 /tmp/mounttest/disk-36.img   0     512
/dev/loop27         0      0         1  0 /tmp/mounttest/disk-26.img   0     512
/dev/loop55         0      0         1  0 /tmp/mounttest/disk-55.img   0     512
/dev/loop17         0      0         1  0 /tmp/mounttest/disk-18.img   0     512
/dev/loop45         0      0         1  0 /tmp/mounttest/disk-45.img   0     512
/dev/loop8          0      0         1  0 /tmp/mounttest/disk-10.img   0     512
/dev/loop35         0      0         1  0 /tmp/mounttest/disk-40.img   0     512
/dev/loop25         0      0         1  0 /tmp/mounttest/disk-27.img   0     512
/dev/loop53         0      0         1  0 /tmp/mounttest/disk-59.img   0     512
/dev/loop15         0      0         1  0 /tmp/mounttest/disk-21.img   0     512
/dev/loop43         0      0         1  0 /tmp/mounttest/disk-44.img   0     512
/dev/loop6          0      0         1  0 /tmp/mounttest/disk-2.img    0     512
/dev/loop33         0      0         1  0 /tmp/mounttest/disk-33.img   0     512
/dev/loop61         0      0         1  0 /tmp/mounttest/disk-62.img   0     512
/dev/loop23         0      0         1  0 /tmp/mounttest/disk-29.img   0     512
/dev/loop51         0      0         1  0 /tmp/mounttest/disk-43.img   0     512
/dev/loop13         0      0         1  0 /tmp/mounttest/disk-16.img   0     512
/dev/loop41         0      0         1  0 /tmp/mounttest/disk-51.img   0     512
/dev/loop4          0      0         1  0 /tmp/mounttest/disk-7.img    0     512
/dev/loop31         0      0         1  0 /tmp/mounttest/disk-32.img   0     512
/dev/loop21         0      0         1  0 /tmp/mounttest/disk-24.img   0     512
/dev/loop11         0      0         1  0 /tmp/mounttest/disk-15.img   0     512
/dev/loop2          0      0         1  0 /tmp/mounttest/disk-23.img   0     512
/dev/loop58         0      0         1  0 /tmp/mounttest/disk-64.img   0     512
/dev/loop48         0      0         1  0 /tmp/mounttest/disk-46.img   0     512
/dev/loop38         0      0         1  0 /tmp/mounttest/disk-42.img   0     512
/dev/loop0          0      0         1  0 /tmp/mounttest/disk-9.img    0     512
/dev/loop28         0      0         1  0 /tmp/mounttest/disk-30.img   0     512
/dev/loop56         0      0         1  0 /tmp/mounttest/disk-61.img   0     512
/dev/loop18         0      0         1  0 /tmp/mounttest/disk-20.img   0     512
/dev/loop46         0      0         1  0 /tmp/mounttest/disk-48.img   0     512
/dev/loop9          0      0         1  0 /tmp/mounttest/disk-12.img   0     512
/dev/loop36         0      0         1  0 /tmp/mounttest/disk-37.img   0     512
/dev/loop26         0      0         1  0 /tmp/mounttest/disk-11.img   0     512
/dev/loop54         0      0         1  0 /tmp/mounttest/disk-53.img   0     512
/dev/loop16         0      0         1  0 /tmp/mounttest/disk-17.img   0     512
/dev/loop44         0      0         1  0 /tmp/mounttest/disk-41.img   0     512
/dev/loop7          0      0         1  0 /tmp/mounttest/disk-6.img    0     512
/dev/loop34         0      0         1  0 /tmp/mounttest/disk-38.img   0     512
/dev/loop24         0      0         1  0 /tmp/mounttest/disk-28.img   0     512
/dev/loop52         0      0         1  0 /tmp/mounttest/disk-57.img   0     512
/dev/loop14         0      0         1  0 /tmp/mounttest/disk-13.img   0     512
/dev/loop42         0      0         1  0 /tmp/mounttest/disk-47.img   0     512
/dev/loop5          0      0         1  0 /tmp/mounttest/disk-3.img    0     512
/dev/loop32         0      0         1  0 /tmp/mounttest/disk-34.img   0     512
/dev/loop60         0      0         1  0 /tmp/mounttest/disk-66.img   0     512
/dev/loop22         0      0         1  0 /tmp/mounttest/disk-99.img   0     512
/dev/loop50         0      0         1  0 /tmp/mounttest/disk-56.img   0     512
/dev/loop12         0      0         1  0 /tmp/mounttest/disk-8.img    0     512
/dev/loop40         0      0         1  0 /tmp/mounttest/disk-39.img   0     512
/dev/loop3          0      0         1  0 /tmp/mounttest/disk-5.img    0     512
/dev/loop30         0      0         1  0 /tmp/mounttest/disk-35.img   0     512
/dev/loop59         0      0         1  0 /tmp/mounttest/disk-54.img   0     512
/dev/loop20         0      0         1  0 /tmp/mounttest/disk-25.img   0     512
/dev/loop49         0      0         1  0 /tmp/mounttest/disk-52.img   0     512
/dev/loop10         0      0         1  0 /tmp/mounttest/disk-14.img   0     512
/dev/loop39         0      0         1  0 /tmp/mounttest/disk-49.img   0     512
```

**Comment by bluca** [comment]
Ran the reproducer script posted at https://github.com/systemd/systemd/files/3524856/reproduce3.sh.log and it failed at iteration 11 on disk 111:

```
# PATH=/tmp:$PATH ./reproduce3.sh.log
/tmp/bc
--- run 1 Mon Sep  2 16:38:53 UTC 2019
--- run 2 Mon Sep  2 16:40:53 UTC 2019
--- run 3 Mon Sep  2 16:42:58 UTC 2019
--- run 4 Mon Sep  2 16:45:23 UTC 2019
--- run 5 Mon Sep  2 16:47:34 UTC 2019
--- run 6 Mon Sep  2 16:49:27 UTC 2019
--- run 7 Mon Sep  2 16:51:35 UTC 2019
--- run 8 Mon Sep  2 16:53:31 UTC 2019
--- run 9 Mon Sep  2 16:55:40 UTC 2019
--- run 10 Mon Sep  2 16:58:09 UTC 2019
--- run 11 Mon Sep  2 17:00:19 UTC 2019
disk-111 missing
mountinfo: 18 0 253:48 / / ro,relatime shared:1 - ext4 /dev/root ro,seclabel,i_version
19 18 0:6 / /dev rw,relatime shared:2 - devtmpfs devtmpfs rw,seclabel,size=508880k,nr_inodes=127220,mode=755
20 18 0:18 / /sys rw,nosuid,nodev,noexec,relatime shared:5 - sysfs sysfs rw,seclabel
21 18 0:4 / /proc rw,relatime shared:10 - proc proc rw
22 20 0:7 / /sys/kernel/security rw,relatime shared:6 - securityfs securityfs rw
23 20 0:17 / /sys/fs/selinux rw,relatime shared:7 - selinuxfs selinuxfs rw
24 19 0:19 / /dev/shm rw,nosuid,nodev shared:3 - tmpfs tmpfs rw,seclabel
25 19 0:20 / /dev/pts rw,relatime shared:4 - devpts devpts rw,seclabel,gid=5,mode=620,ptmxmode=000
26 18 0:21 / /run rw,nosuid,nodev shared:11 - tmpfs tmpfs rw,seclabel,mode=755
27 20 0:22 / /sys/fs/cgroup rw,nosuid,nodev,noexec,relatime shared:8 - cgroup2 cgroup2 rw,seclabel,nsdelegate
28 20 0:23 / /sys/fs/bpf rw,nosuid,nodev,noexec,relatime shared:9 - bpf bpf rw,mode=700
29 19 0:24 / /dev/hugepages rw,relatime shared:12 - hugetlbfs hugetlbfs rw,seclabel,pagesize=2M
30 18 0:25 / /tmp rw,nosuid,nodev shared:13 - tmpfs tmpfs rw,rootcontext=system_u:object_r:tmp_t,seclabel
31 20 0:8 / /sys/kernel/debug rw,nosuid,nodev,noexec,relatime shared:14 - debugfs debugfs rw,seclabel
32 19 0:16 / /dev/mqueue rw,nosuid,nodev,noexec,relatime shared:15 - mqueue mqueue rw,seclabel
66 18 0:26 / /var/volatile rw,relatime shared:33 - tmpfs tmpfs rw,rootcontext=system_u:object_r:var_t,seclabel
69 18 0:26 /log/utmp /var/log/utmp rw,relatime shared:33 - tmpfs tmpfs rw,rootcontext=system_u:object_r:var_t,seclabel
72 18 0:26 /log/lastlog /var/log/lastlog rw,relatime shared:33 - tmpfs tmpfs rw,rootcontext=system_u:object_r:var_t,seclabel
75 18 0:26 /log/wtmp /var/log/wtmp rw,relatime shared:33 - tmpfs tmpfs rw,rootcontext=system_u:object_r:var_t,seclabel
78 18 0:26 /log/btmp /var/log/btmp rw,relatime shared:33 - tmpfs tmpfs rw,rootcontext=system_u:object_r:var_t,seclabel
83 18 0:28 / /var/cache rw,relatime shared:39 - overlay overlay rw,seclabel,lowerdir=/var/cache,upperdir=/var/volatile/cache,workdir=/var/volatile/.cache-work
81 18 0:27 / /var/spool rw,relatime shared:41 - overlay overlay rw,seclabel,lowerdir=/var/spool,upperdir=/var/volatile/spool,workdir=/var/volatile/.spool-work
88 18 0:30 / /var/lib rw,relatime shared:43 - overlay overlay rw,seclabel,lowerdir=/var/lib,upperdir=/var/volatile/lib,workdir=/var/volatile/.lib-work
96 18 253:0 / /vol/data rw,relatime shared:45 - ext4 /dev/vda rw,seclabel
99 18 253:0 /var/log/journal /var/log/journal rw,relatime shared:45 - ext4 /dev/vda rw,seclabel
413 26 0:35 / /run/user/0 rw,nosuid,nodev,relatime shared:216 - tmpfs tmpfs rw,seclabel,size=101876k,mode=700
457 31 0:11 / /sys/kernel/debug/tracing rw,nosuid,nodev,noexec,relatime shared:237 - tracefs tracefs rw,seclabel
978 30 7:59 / /tmp/mounttest/mount/61 rw,nodev,relatime shared:549 - ext4 /dev/loop59 rw,seclabel
1041 30 7:62 / /tmp/mounttest/mount/63 rw,nodev,relatime shared:567 - ext4 /dev/loop62 rw,seclabel
1063 30 7:64 / /tmp/mounttest/mount/72 rw,nodev,relatime shared:195 - ext4 /dev/loop64 rw,seclabel
1062 30 7:63 / /tmp/mounttest/mount/67 rw,nodev,relatime shared:231 - ext4 /dev/loop63 rw,seclabel
1064 30 7:65 / /tmp/mounttest/mount/65 rw,nodev,relatime shared:183 - ext4 /dev/loop65 rw,seclabel
362 30 7:66 / /tmp/mounttest/mount/70 rw,nodev,relatime shared:189 - ext4 /dev/loop66 rw,seclabel
363 30 7:70 / /tmp/mounttest/mount/68 rw,nodev,relatime shared:213 - ext4 /dev/loop70 rw,seclabel
365 30 7:68 / /tmp/mounttest/mount/75 rw,nodev,relatime shared:225 - ext4 /dev/loop68 rw,seclabel
366 30 7:67 / /tmp/mounttest/mount/59 rw,nodev,relatime shared:243 - ext4 /dev/loop67 rw,seclabel
408 30 7:69 / /tmp/mounttest/mount/69 rw,nodev,relatime shared:249 - ext4 /dev/loop69 rw,seclabel
319 30 7:71 / /tmp/mounttest/mount/73 rw,nodev,relatime shared:573 - ext4 /dev/loop71 rw,seclabel
431 30 7:72 / /tmp/mounttest/mount/71 rw,nodev,relatime shared:207 - ext4 /dev/loop72 rw,seclabel
432 30 7:73 / /tmp/mounttest/mount/79 rw,nodev,relatime shared:579 - ext4 /dev/loop73 rw,seclabel
352 30 7:75 / /tmp/mounttest/mount/87 rw,nodev,relatime shared:201 - ext4 /dev/loop75 rw,seclabel
479 30 7:74 / /tmp/mounttest/mount/76 rw,nodev,relatime shared:255 - ext4 /dev/loop74 rw,seclabel
385 30 7:76 / /tmp/mounttest/mount/66 rw,nodev,relatime shared:261 - ext4 /dev/loop76 rw,seclabel
386 30 7:78 / /tmp/mounttest/mount/74 rw,nodev,relatime shared:267 - ext4 /dev/loop78 rw,seclabel
388 30 7:77 / /tmp/mounttest/mount/78 rw,nodev,relatime shared:285 - ext4 /dev/loop77 rw,seclabel
387 30 7:79 / /tmp/mounttest/mount/77 rw,nodev,relatime shared:585 - ext4 /dev/loop79 rw,seclabel
394 30 7:6 / /tmp/mounttest/mount/80 rw,nodev,relatime shared:591 - ext4 /dev/loop6 rw,seclabel
495 30 7:2 / /tmp/mounttest/mount/85 rw,nodev,relatime shared:297 - ext4 /dev/loop2 rw,seclabel
1108 30 7:0 / /tmp/mounttest/mount/84 rw,nodev,relatime shared:279 - ext4 /dev/loop0 rw,seclabel
1109 30 7:3 / /tmp/mounttest/mount/82 rw,nodev,relatime shared:315 - ext4 /dev/loop3 rw,seclabel
528 30 7:7 / /tmp/mounttest/mount/89 rw,nodev,relatime shared:291 - ext4 /dev/loop7 rw,seclabel
523 30 7:1 / /tmp/mounttest/mount/83 rw,nodev,relatime shared:309 - ext4 /dev/loop1 rw,seclabel
592 30 7:4 / /tmp/mounttest/mount/92 rw,nodev,relatime shared:321 - ext4 /dev/loop4 rw,seclabel
529 30 7:8 / /tmp/mounttest/mount/81 rw,nodev,relatime shared:327 - ext4 /dev/loop8 rw,seclabel
524 30 7:9 / /tmp/mounttest/mount/90 rw,nodev,relatime shared:303 - ext4 /dev/loop9 rw,seclabel
578 30 7:10 / /tmp/mounttest/mount/86 rw,nodev,relatime shared:273 - ext4 /dev/loop10 rw,seclabel
579 30 7:80 / /tmp/mounttest/mount/91 rw,nodev,relatime shared:333 - ext4 /dev/loop80 rw,seclabel
581 30 7:5 / /tmp/mounttest/mount/88 rw,nodev,relatime shared:339 - ext4 /dev/loop5 rw,seclabel
492 30 7:13 / /tmp/mounttest/mount/96 rw,nodev,relatime shared:345 - ext4 /dev/loop13 rw,seclabel
625 30 7:11 / /tmp/mounttest/mount/97 rw,nodev,relatime shared:351 - ext4 /dev/loop11 rw,seclabel
626 30 7:81 / /tmp/mounttest/mount/93 rw,nodev,relatime shared:363 - ext4 /dev/loop81 rw,seclabel
580 30 7:82 / /tmp/mounttest/mount/94 rw,nodev,relatime shared:369 - ext4 /dev/loop82 rw,seclabel
686 30 7:83 / /tmp/mounttest/mount/95 rw,nodev,relatime shared:375 - ext4 /dev/loop83 rw,seclabel
687 30 7:14 / /tmp/mounttest/mount/100 rw,nodev,relatime shared:381 - ext4 /dev/loop14 rw,seclabel
698 30 7:15 / /tmp/mounttest/mount/98 rw,nodev,relatime shared:387 - ext4 /dev/loop15 rw,seclabel
668 30 7:17 / /tmp/mounttest/mount/107 rw,nodev,relatime shared:357 - ext4 /dev/loop17 rw,seclabel
646 30 7:16 / /tmp/mounttest/mount/102 rw,nodev,relatime shared:399 - ext4 /dev/loop16 rw,seclabel
670 30 7:19 / /tmp/mounttest/mount/99 rw,nodev,relatime shared:405 - ext4 /dev/loop19 rw,seclabel
754 30 7:12 / /tmp/mounttest/mount/110 rw,nodev,relatime shared:411 - ext4 /dev/loop12 rw,seclabel
755 30 7:22 / /tmp/mounttest/mount/109 rw,nodev,relatime shared:417 - ext4 /dev/loop22 rw,seclabel
671 30 7:20 / /tmp/mounttest/mount/114 rw,nodev,relatime shared:423 - ext4 /dev/loop20 rw,seclabel
669 30 7:21 / /tmp/mounttest/mount/105 rw,nodev,relatime shared:429 - ext4 /dev/loop21 rw,seclabel
756 30 7:18 / /tmp/mounttest/mount/103 rw,nodev,relatime shared:435 - ext4 /dev/loop18 rw,seclabel
757 30 7:23 / /tmp/mounttest/mount/101 rw,nodev,relatime shared:597 - ext4 /dev/loop23 rw,seclabel
758 30 7:25 / /tmp/mounttest/mount/104 rw,nodev,relatime shared:603 - ext4 /dev/loop25 rw,seclabel
677 30 7:24 / /tmp/mounttest/mount/106 rw,nodev,relatime shared:393 - ext4 /dev/loop24 rw,seclabel
813 30 7:28 / /tmp/mounttest/mount/113 rw,nodev,relatime shared:609 - ext4 /dev/loop28 rw,seclabel
820 30 7:29 / /tmp/mounttest/mount/112 rw,nodev,relatime shared:615 - ext4 /dev/loop29 rw,seclabel
1123 30 7:30 / /tmp/mounttest/mount/115 rw,nodev,relatime shared:621 - ext4 /dev/loop30 rw,seclabel
1142 30 7:26 / /tmp/mounttest/mount/108 rw,nodev,relatime shared:627 - ext4 /dev/loop26 rw,seclabel
1163 30 7:31 / /tmp/mounttest/mount/116 rw,nodev,relatime shared:633 - ext4 /dev/loop31 rw,seclabel
745 30 7:32 / /tmp/mounttest/mount/119 rw,nodev,relatime shared:447 - ext4 /dev/loop32 rw,seclabel
1195 30 7:27 / /tmp/mounttest/mount/120 rw,nodev,relatime shared:453 - ext4 /dev/loop27 rw,seclabel
843 30 7:35 / /tmp/mounttest/mount/118 rw,nodev,relatime shared:465 - ext4 /dev/loop35 rw,seclabel
832 30 7:36 / /tmp/mounttest/mount/117 rw,nodev,relatime shared:471 - ext4 /dev/loop36 rw,seclabel
-- failed units, skipping cleanup, use --cleanup manually
```

**Comment by karelzak** [comment]
@bluca 
> EDIT: a second run completed successfully all 40 iterations, while with ~20 units it's regularly reproducible.

Well, I guess it's original reproducer with "cat /proc/self/mountinfo" in veryfy() function, right?

It's expected that it will fail as cat(1) does not use read()+poll() to read mountinfo. You need to use findmnt or for example "./test_mount_utils --read-procfs /proc/self/mountinfo" from util-linux git tree in the test instead of cat(1).

The reproducer is also not about systemd. It's about kernel and userspace mountinfo races only.

**Comment by bluca** [comment]
> @bluca
> 
> > EDIT: a second run completed successfully all 40 iterations, while with ~20 units it's regularly reproducible.
> 
> Well, I guess it's original reproducer with "cat /proc/self/mountinfo" in verify() function, right?
> 
> It's expected that it will fail as cat(1) does not use read()+poll() to read mountinfo. You need to use findmnt or for example "./test_mount_utils --read-procfs /proc/self/mountinfo" from util-linux git tree in the test instead of cat(1).

Thank you, will try that later as well.

> The reproducer is also not about systemd. It's about kernel and userspace mountinfo races only.

Yep - I tried both to try and see if I was missing something on the libmount side. Thanks!

**Comment by karelzak** [comment]
And note that @keszybz has created the reproducer to show that the problem is not systemd, but the way how kernel exports mount table to userspace and how we (libmount, systemd, cat, ...whatever) read it. Unfortunately, it's probably not possible have a 100% fix for this issue in userspace -- let's hope that next kernel versions will be better (and maybe a new fsinfo() kernel syscall will help us too). 

For now the goal is to make the race less visible (less frequented) :-)

**Comment by bluca** [comment]
Yep makes sense - although now that systemd uses libmount internally (with some of the patches mentioned earlier) it should in theory get better, right?

Is there already a kernel patch to try? I'm happy to test stuff even if half done!

**Comment by bboozzoo** [comment]
Thank you everyone for digging into this!

**Comment by karelzak** [comment]
> Is there already a kernel patch to try?
Oh, man you're an real optimist :-)

**Comment by karelzak** [comment]
Note that Fedora rawhide is now with the libmount improvement. So, it should be relatively easy to test it the original use-case.

**Comment by mikkorapeli** [comment]
If this is a kernel side race, then it affects a lot of kernel versions. I've been affected with 4.14. Can't run systemd 237 and later without patch like https://github.com/systemd/systemd/issues/10872#issuecomment-469555708 which fixes unreliable mounts on several different HW setups. Bug report link to kernel bugzilla would be nice to have. Unfortunately there is no documentation which says that /proc/${PID}/mountinfo is uptodate after mount command succeeds.

**Comment by keszybz** [comment]
I can still reproduce the issue using ./reproduce3.sh with util-linux-2.34-6.fc32.x86_64 ;(

**Comment by karelzak** [comment]
> I can still reproduce the issue using ./reproduce3.sh with util-linux-2.34-6.fc32.x86_64 ;(

I guess there is "cat /proc/self/mountinfo" in the test ... you need findmnt or so. cat(1) does not implement the workaround :-) Try modify verify() and use:
```
    findmnt --noheadings --mountpoint "$PREFIX/mount/$i" >/dev/null
    if [ $? != 0 ]; then
        echo "disk-$i missing"
        return 1
    fi
```

**Comment by keszybz** [comment]
Duh. Sorry, I got confused.

**Comment by keszybz** [comment]
Good news: the issue with /proc/self/mountinfo seems to be fixed, in the sense that `mount` always reports the expected mounts to be present. I cannot reproduce the issue with reproduce3.sh.log from https://github.com/systemd/systemd/issues/10872#issuecomment-523399087.

But variant0 from the OP still fails.

I'm using the original reproduce from OP, variant0, with `cat /proc/self/mountinfo` replaced with `findmnt`. It seems systemd gets confused as to what is mounted, maybe because it is asynchronously receiving information about device being unmounted:
```
--- run 1 Fri Oct 11 16:36:37 CEST 2019
+ systemctl start tmp-mounttest-mount-4.mount
+ systemctl start tmp-mounttest-mount-7.mount
4: /dev/loop0
+ systemctl stop tmp-mounttest-mount-4.mount
+ systemctl start tmp-mounttest-mount-14.mount
+ systemctl start tmp-mounttest-mount-15.mount
+ systemctl start tmp-mounttest-mount-1.mount
+ systemctl start tmp-mounttest-mount-8.mount
+ systemctl start tmp-mounttest-mount-9.mount
+ systemctl start tmp-mounttest-mount-10.mount
+ systemctl start tmp-mounttest-mount-16.mount
+ systemctl start tmp-mounttest-mount-5.mount
+ systemctl start tmp-mounttest-mount-13.mount
+ systemctl start tmp-mounttest-mount-12.mount
10: /dev/loop6
+ systemctl stop tmp-mounttest-mount-10.mount
14: /dev/loop0
+ systemctl stop tmp-mounttest-mount-14.mount
8: /dev/loop4
+ systemctl stop tmp-mounttest-mount-8.mount
+ systemctl start tmp-mounttest-mount-11.mount
+ systemctl start tmp-mounttest-mount-2.mount
+ systemctl start tmp-mounttest-mount-17.mount
7: /dev/loop1
+ systemctl stop tmp-mounttest-mount-7.mount
+ systemctl start tmp-mounttest-mount-3.mount
+ systemctl start tmp-mounttest-mount-20.mount
11: /dev/loop0 <-----------------------------------------------
+ systemctl stop tmp-mounttest-mount-11.mount
+ systemctl start tmp-mounttest-mount-18.mount
1: /dev/loop3
+ systemctl stop tmp-mounttest-mount-1.mount
+ systemctl start tmp-mounttest-mount-6.mount
+ systemctl start tmp-mounttest-mount-19.mount
5: /dev/loop8
+ systemctl stop tmp-mounttest-mount-5.mount
16: /dev/loop7
+ systemctl stop tmp-mounttest-mount-16.mount
6: /dev/loop3
+ systemctl stop tmp-mounttest-mount-6.mount
13: /dev/loop9
+ systemctl stop tmp-mounttest-mount-13.mount
15: /dev/loop2
+ systemctl stop tmp-mounttest-mount-15.mount
17: /dev/loop6
+ systemctl stop tmp-mounttest-mount-17.mount
18: /dev/loop0
+ systemctl stop tmp-mounttest-mount-18.mount
9: /dev/loop5
+ systemctl stop tmp-mounttest-mount-9.mount
disk-20 missing
* tmp-mounttest-mount-20.mount - /tmp/mounttest/mount/20
   Loaded: loaded (/etc/systemd/system/tmp-mounttest-mount-20.mount; static; vendor preset: disabled)
   Active: inactive (dead)
    Where: /tmp/mounttest/mount/20
     What: /tmp/mounttest/disk-20.img

Oct 11 16:36:42 workstation-uefi systemd[1]: tmp-mounttest-mount-20.mount: tmp-mounttest-mount-20.mount lost dependency BindsTo=dev-loop0.device
Oct 11 16:36:42 workstation-uefi systemd[1]: tmp-mounttest-mount-20.mount: tmp-mounttest-mount-20.mount lost dependency After=dev-loop0.device
Oct 11 16:36:42 workstation-uefi systemd[1]: tmp-mounttest-mount-20.mount: tmp-mounttest-mount-20.mount lost dependency References=dev-loop0.device
Oct 11 16:36:42 workstation-uefi systemd[1]: tmp-mounttest-mount-20.mount: Collecting.
```

**Comment by nomuranec** [comment]
I'm hitting the same problem and can reproduce it by running this:
`   while systemctl start test.mount; do systemctl stop test.mount; sleep 3; done`
and this:
`   while true; do systemctl daemon-reload; done`
in parallel.

As far as I can see, the problem comes out when the unit state is restored to 'mounting' by the coldplug upon daemon reload, then later check finds corresponding entry in /proc/self/mountinfo and notice:
>   Mount process finished, but there is no mount.
>   Failed with result 'protocol'.

How about setting the unit flag 'just_mounted' if the unit is from /proc/self/mountinfo and the state is 'mounting'? A patch like this works for me:

```
diff --git a/src/core/mount.c b/src/core/mount.c
index 6e42176..6a109b6 100644
--- a/src/core/mount.c
+++ b/src/core/mount.c
@@ -1550,7 +1550,7 @@ static int mount_setup_existing_unit(
         if (r > 0)
                 flags |= MOUNT_PROC_JUST_CHANGED;
 
-        if (!MOUNT(u)->from_proc_self_mountinfo || FLAGS_SET(MOUNT(u)->proc_flags, MOUNT_PROC_JUST_MOUNTED))
+        if (!MOUNT(u)->from_proc_self_mountinfo || FLAGS_SET(MOUNT(u)->proc_flags, MOUNT_PROC_JUST_MOUNTED) || MOUNT(u)->state == MOUNT_MOUNTING)
                 flags |= MOUNT_PROC_JUST_MOUNTED;
 
         MOUNT(u)->from_proc_self_mountinfo = true;
```

**Comment by poettering** [comment]
@bboozzoo @keszybz hmm, so, I merged #14234 yesterday, which closes this bug. I'd like to verify whether everything is actually fixed here. Any chance you can re-run those tests with an updated script and current util-linux, to verify that everything is really and actually fixed now?

**Comment by keszybz** [comment]
I cannot reproduce the issue any more as described in https://github.com/systemd/systemd/issues/10872#issuecomment-541114951.

**Comment by bboozzoo** [comment]
@keszybz did you try the scenario with systemctl daemon-reload/start/stop? I've built mkosi Arch image, with systemd f6160131e7 and can see the interleaved start/stop/reload to fail quite consistently. The reproducer script was updated to use findmnt https://gist.github.com/bboozzoo/d4b142229b1915ef7cc0cf8593599ad9 per https://github.com/systemd/systemd/issues/10872#issuecomment-541114951

**Comment by keszybz** [comment]
> did you try the scenario with systemctl daemon-reload/start/stop?

Nope.

> https://gist.github.com/bboozzoo/d4b142229b1915ef7cc0cf8593599ad9 

It fails with this version for me too.

**Comment by nomuranec** [comment]
@bboozzoo @keszybz Do you still see mount failure like "Mount process finished, but there is no mount."?

**Comment by bboozzoo** [comment]
@nomuranec there's no message, but the mount isn't there either.

**Comment by keszybz** [comment]
It fails with "disk-nn missing".

**Comment by nomuranec** [comment]
@bboozzoo @keszybz  thanks, I think I can see the same phemenon with you using https://gist.github.com/bboozzoo/d4b142229b1915ef7cc0cf8593599ad9 .

What's strange is when I put "date" command before "disk-nn missing" message in the reproducer script to see when it checks mountinfo, failure looks like this:

> --- run 8 Fri Jan 17 02:23:30 EST 2020
--- run 9 Fri Jan 17 02:23:44 EST 2020
Fri Jan 17 02:23:57 EST 2020
disk-2 missing

So the 9th loop was started at 02:23:44 and the script checked mountinfo and found an entry was not there at 02:23:57.

Whereas journal for tmp-mounttest-mount-2.mount looks:

> Jan 17 02:23:37 localhost systemd[42214]: tmp-mounttest-mount-2.mount: Executing: /usr/bin/mount /tmp/mounttest/disk-2.img /tmp/mounttest/mount/2 -t ext4 -o nodev
Jan 17 02:23:37 localhost systemd[1]: Mounting mount unit for test 2...
Jan 17 02:23:38 localhost systemd[1]: Mounted mount unit for test 2.
Jan 17 02:23:41 localhost systemd[42738]: tmp-mounttest-mount-2.mount: Executing: /usr/bin/umount /tmp/mounttest/mount/2 -c
Jan 17 02:23:41 localhost systemd[1]: Unmounting mount unit for test 2...
Jan 17 02:23:42 localhost systemd[1]: tmp-mounttest-mount-2.mount: Succeeded.
Jan 17 02:23:42 localhost systemd[1]: Unmounted mount unit for test 2.
Jan 17 02:23:53 localhost systemd[43797]: tmp-mounttest-mount-2.mount: Executing: /usr/bin/mount /tmp/mounttest/disk-2.img /tmp/mounttest/mount/2 -t ext4 -o nodev
Jan 17 02:23:53 localhost systemd[1]: Mounting mount unit for test 2...
Jan 17 02:23:53 localhost systemd[1]: Mounted mount unit for test 2.
Jan 17 02:23:55 localhost systemd[44100]: tmp-mounttest-mount-2.mount: Executing: /usr/bin/umount /tmp/mounttest/mount/2 -c
Jan 17 02:23:55 localhost systemd[1]: Unmounting mount unit for test 2...
Jan 17 02:23:55 localhost systemd[1]: tmp-mounttest-mount-2.mount: Succeeded.
Jan 17 02:23:55 localhost systemd[1]: Unmounted mount unit for test 2.

Before the script tried to verify mountinfo at 02:23:57, systemd had already unmounted at 02:23:55!  It's quite natural that the script couldn't find mount entry.  So the problem seems why systemd did unmount automatically...

**Comment by nomuranec** [comment]
_BTW, I think the latest problem (mount unit is stopped with unknown reason) is different from the original problem (mount unit fails to start with "Mount process finished, but there is no mount."). So it might be better to open a separate issue for the new one to avoid confusing people..._


I observed that when "disk-X missing" happens, there are following order of events:

1. (suppose /dev/loop0 is attached to disk-Y.img for mount-Y.mount)
2. a job is queued to stop mount-Y.mount by "systemctl stop mount-Y".
3. a job is queued to start mount-X.mount by "systemctl start mount-X".
4. umount is done for mount-Y and /dev/loop0 is detached.
5. mount is done for mount-X and /dev/loop0 is attached to disk-X.img now.
6. systemd addds "BindsTo" dependency from mount-X to /dev/loop0.
7. mount-Y.mount is stopped. mount-X.mount is started.
8. a job is queued to stop mount-X.mount by **something**. This is the problem.

Couldn't there be a race between starting mount unit and jobs related to detached /dev/loop, which triggers the unexpected stop of mount unit because of the BindsTo dependency by re-attached /dev/loop?

**Comment by karelzak** [comment]
Note that for the original issue I have used (during work on libmount) "mount -t tmpfs" rather than loopdev to reproduce the kernel issue and to keep systemd and another stuff out of the game. Maybe you can update the reproducer script and add another --variant to reproduce core of the problem. Just idea :-)

**Comment by nomuranec** [comment]
@karelzak thanks, that's good idea.  Here is a modified reproducer: https://gist.github.com/nomuranec/359a4495900f26a1befa4380634b130f

I added variant 5 (--variant-static-lodev), which does same thing as variant 0 except that relationship between image file and loopback device doesn't change during running.  That is:
- variant 0 specifies a path to image file by "What=". mount/umount command automatically attaches/detaches loopback device to the file.
- variant 5 attaches loopback device for each image file before starting the loops and specifies the device node by "What=".

With variant 0, "disk-nn missing" phenomenon still reproduces reliably within 40 loops with today's head e1d32d6ee86aa.
With variant 5, it doesn't even when I increase $LOOPS to 400.

**Comment by keszybz** [comment]
We should return to this at some point...

**Comment by bluca** [comment]
> We should return to this at some point...

Would the fsinfo syscal proposed at https://lore.kernel.org/lkml/1445647.1585576702@warthog.procyon.org.uk/ help with this race condition?

**Comment by karelzak** [comment]
> Would the fsinfo syscal proposed at https://lore.kernel.org/lkml/1445647.1585576702@warthog.procyon.org.uk/ help with this race condition?

I'm working with David Howells and  Ian Kent on this, but it will take some time as the patch and the interface is still under development. The goal is to hide all the details to libmount. For systemd it will be necessary to use a new mount notification interface (not in kernel yet) which is based on mount IDs.

**Comment by syyhao1994** [comment]
> It fails with "disk-nn missing".

I encounterd the same problem, so this problem is still unreasolvable?

**Comment by raven-au** [comment]
> > It fails with "disk-nn missing".
> 
> I encounterd the same problem, so this problem is still unreasolvable?

What kernel version are you using?

**Comment by syyhao1994** [comment]
> > > It fails with "disk-nn missing".
> > 
> > 
> > I encounterd the same problem, so this problem is still unreasolvable?
> 
> What kernel version are you using?

4.19.
And i found the problem that when i run the reproducer https://gist.github.com/nomuranec/359a4495900f26a1befa4380634b130f
on aarch64 with systemd 239 or 243, the error is "Mount process finished, but there is no mount."
but on x86_64 the error is disk-X missing with variant_0.

**Comment by syyhao1994** [comment]
> "disk-nn missing"

So maybe the "disk-nn missing" phenomenon is caused by the reproducer rather than a problem?

**Comment by raven-au** [comment]
> > "disk-nn missing"
> 
> So maybe the "disk-nn missing" phenomenon is caused by the reproducer rather than a problem?
> @nomuranec

It's my understanding this is caused by a locking problem in the kernel when reading the proc mounts table.
There was also a temporary work around added to libmount (of util-linux).

The kernel fix went into v5.7 or v5.8, I'm not sure what version of util-linux has the workaround.

So if your libmount doesn't have the workaround or your distribution kernel (or the kernel you are building) doesn't have the fix you will see the problem.

**Comment by karelzak** [comment]
> The kernel fix went into v5.7 or v5.8, I'm not sure what version of util-linux has the workaround.

util-linux v2.35 and v2.36  (the current upstream (not released yet) is without the workaround)

**Comment by syyhao1994** [comment]
> > > "disk-nn missing"
> > 
> > 
> > So maybe the "disk-nn missing" phenomenon is caused by the reproducer rather than a problem?
> > @nomuranec
> 
> It's my understanding this is caused by a locking problem in the kernel when reading the proc mounts table.
> There was also a temporary work around added to libmount (of util-linux).
> 
> The kernel fix went into v5.7 or v5.8, I'm not sure what version of util-linux has the workaround.
> 
> So if your libmount doesn't have the workaround or your distribution kernel (or the kernel you are building) doesn't have the fix you will see the problem.

I try the workaround patch in util-linux of https://github.com/karelzak/util-linux/commit/e4925f591c1bfb83719418b56b952830d15b77eb.
But as we could see through runnin the producer, the "disk-nn missing" is still there.

**Comment by karelzak** [comment]
@syyhao1994 there are more patches related to this topic. You need also ee551c909f95437fd9fcd162f398c069d0ce9720.

**Comment by syyhao1994** [comment]
> @syyhao1994 there are more patches related to this topic. You need also ee551c909f95437fd9fcd162f398c069d0ce9720.

You mean the problem of "disk-xx missing"?

**Comment by nomuranec** [comment]
@syyhao1994 Have you tried variant 5 (`--variant-static-lodev`)of the test program?  Variant 5 is simplified version of variant 0.  The mappings of lodev and disk-xx.img are static in variant 5 during the test instead of dynamic as in variant 0.  IMO this issue is complicated by involving not only mount but also lodev mapping changes.

**Comment by syyhao1994** [comment]
> @karelzak thanks, that's good idea. Here is a modified reproducer: https://gist.github.com/nomuranec/359a4495900f26a1befa4380634b130f
> 
> I added variant 5 (--variant-static-lodev), which does same thing as variant 0 except that relationship between image file and loopback device doesn't change during running. That is:
> 
> * variant 0 specifies a path to image file by "What=". mount/umount command automatically attaches/detaches loopback device to the file.
> * variant 5 attaches loopback device for each image file before starting the loops and specifies the device node by "What=".
> 
> With variant 0, "disk-nn missing" phenomenon still reproduces reliably within 40 loops with today's head [e1d32d6](https://github.com/systemd/systemd/commit/e1d32d6ee86aa0515b4aeac0ff8ee89f11f0f45e).
> With variant 5, it doesn't even when I increase $LOOPS to 400.

@karelzak As nomuranec said, the problem is still exist with variant 0

**Comment by syyhao1994** [comment]
> @syyhao1994 Have you tried variant 5 (`--variant-static-lodev`)of the test program? Variant 5 is simplified version of variant 0. The mappings of lodev and disk-xx.img are static in variant 5 during the test instead of dynamic as in variant 0. IMO this issue is complicated by involving not only mount but also lodev mapping changes.

I tried many times, the problem is still exist with variant 5, but it was very hard to reproduce， i have tried maybe thousands of time.
With variant 0  it is very easy to reproduce.
I am so confused with this phenomenon.

**Comment by nomuranec** [comment]
Because variant 0 has high frequency lodev mapping change in the mix, it could trigger other problem than /proc/self/mountinfo race.  If you are chasing mount-related problem, variant 5 has tighter focus on that.  If variant 0 matches your use case, IMHO forking a new issue with specific systemd/kernel/util-linux version avoids confusion...

**Comment by syyhao1994** [comment]
> Because variant 0 has high frequency lodev mapping change in the mix, it could trigger other problem than /proc/self/mountinfo race. If you are chasing mount-related problem, variant 5 has tighter focus on that. If variant 0 matches your use case, IMHO forking a new issue with specific systemd/kernel/util-linux version avoids confusion...

Ok, it may be couldn't solve in systemd. Thank you a lot!

**Comment by karelzak** [comment]
Frankly, I do not see reason why use loopdev to test mountinfo issues. It only inceases complexity and kernel loopdev driver is pretty problematic when used in paralell. It seesm better to minimize complexity and use for example "tmpfs" to test mountinfo.

See for example my version of the script which I have used to implement the workaround: http://people.redhat.com/kzak/rep-tmpfs.sh.

Anyway, the mountinfo read() issue should be fixed by kernel (since 5.8).

**Comment by yuwata** [comment]
Hopefully fixed by #23893.
